File: packages/shared-types/src/cv.ts
import { z } from "zod";

export const LinkSchema = z.object({
  name: z.string(),
  href: z.string().url("Must be a valid URL (include https://)"),
});

export const PersonalInfoSchema = z.object({
  fullName: z.string(),
  email: z.string().email("Must be a valid email address"),
  phone: z
    .string()
    .optional()
    .refine(
      (v) => !v || /^\+?[1-9]\d{6,14}$/.test(v.replace(/[\s\-().]/g, "")),
      "Must be a valid phone number (e.g. +44 7700 900000)",
    ),
  location: z.string().optional(),
  title: z.string().optional(),
  summary: z.string().optional(),
  links: z.array(LinkSchema).optional(),
});

export const ExperienceSchema = z.object({
  id: z.string(),
  company: z.string(),
  title: z.string(),
  location: z.string().optional(),
  startDate: z.string(),
  endDate: z.string().optional(),
  current: z.boolean().optional(),
  description: z.string().optional(),
  highlights: z.array(z.string()).optional(),
});

export const EducationSchema = z.object({
  id: z.string(),
  institution: z.string(),
  degree: z.string(),
  field: z.string(),
  startDate: z.string(),
  endDate: z.string().optional(),
  gpa: z.string().optional(),
});

export const ProjectSchema = z.object({
  id: z.string(),
  name: z.string(),
  role: z.string().optional(),
  startDate: z.string().optional(),
  endDate: z.string().optional(),
  current: z.boolean().optional(),
  description: z.string().optional(),
  technologies: z.array(z.string()).optional(),
});

export const CertificationSchema = z.object({
  id: z.string(),
  name: z.string(),
  issuer: z.string().optional(),
  date: z.string().optional(),
});

export const CvSchema = z.object({
  id: z.string(),
  userId: z.string(),
  fileName: z.string().optional(),
  version: z.number().optional(),
  isActive: z.boolean().optional(),
  personalInfo: PersonalInfoSchema,
  experience: z.array(ExperienceSchema),
  education: z.array(EducationSchema),
  skills: z.array(z.string()),
  projects: z.array(ProjectSchema).optional(),
  certifications: z.array(CertificationSchema).optional(),
  extraCurricular: z.array(z.string()).optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  normalized: z.any().optional(),
});

export type CvLink = z.infer<typeof LinkSchema>;
export type PersonalInfo = z.infer<typeof PersonalInfoSchema>;
export type Experience = z.infer<typeof ExperienceSchema>;
export type Education = z.infer<typeof EducationSchema>;
export type Project = z.infer<typeof ProjectSchema>;
export type Certification = z.infer<typeof CertificationSchema>;
export type Cv = z.infer<typeof CvSchema>;

export type CV = Cv;

File: apps/ai-service/app/schemas/cv.py
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field


class PersonalInfo(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    fullName: str = ""  # noqa: N815
    email: str = ""
    phone: str | None = None
    location: str | None = None
    title: str | None = None
    summary: str | None = None


class Experience(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(default_factory=lambda: str(uuid4()))
    company: str
    title: str
    location: str | None = None
    startDate: str | None = None  # noqa: N815
    endDate: str | None = None  # noqa: N815
    current: bool | None = None
    description: str | None = None
    highlights: list[str] = []


class Education(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(default_factory=lambda: str(uuid4()))
    institution: str
    degree: str
    field: str
    startDate: str | None = None  # noqa: N815
    endDate: str | None = None  # noqa: N815
    gpa: str | None = None


class ParsedCvResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    personalInfo: PersonalInfo  # noqa: N815
    experience: list[Experience] = []
    education: list[Education] = []
    skills: list[str] = []
    fileName: str = ""  # noqa: N815
